### ISSUE TITLE
Make IStartup and IStartupFilter async

### ISSUE THREAD

**Issue by anurse** [aspnet/Hosting#1088]
`IStartup` (and this Startup classes) and `IStartupFilter` should be async. If Start/Stop on the WebHost are async, we should flow the async-ness down

**Comment by Drawaes** [comment]
Please do! I could use this now

**Comment by davidfowl** [comment]
Don't change it, just add a new interface.

**Comment by anurse** [comment]
That's basically what I meant, updating to clarify.

**Comment by khellang** [comment]
Why does it have to be an error to have both? Couldn't you just have async be prioritized over sync? I.e. resolve `IAsyncStartup` first, then if It's not registered, resolve `IStartup` etc. `StartupBase` could implement both and have the async overload call the sync overload.

**Comment by davidfowl** [comment]
I think async should win over sync

**Comment by anurse** [comment]
We certainly could, it just seems like it's unnecessarily ambiguous. Wouldn't you have to explicitly register both to get in that scenario? All our helpers should be updated to register just the one (`IAsyncStartup`) with the necessary adaptors.

**Comment by davidfowl** [comment]
@anurse you may implement both on the same object.

**Comment by anurse** [comment]
I suppose. I just still think that's a pretty unclear situation, since you'll have code that we'll never call ourselves. Anyway, prioritizing async is the right approach if we're going to allow both.

**Comment by Tratcher** [comment]
@davidfowl The call stack for this doesn't make sense. Startup is instantiated and run from WebHostBuilder.Build(), not WebHost.StartAsync(), so Build() would have to become async for this.

**Comment by anurse** [comment]
Yeah. Making Build async is out of scope for 2.0 I think, it's way too late. What about reopening #1085 and moving `IHostedService` initialization to before Server initialization?

**Comment by Tratcher** [comment]
@muratg discussed this with @davidfowl offline, this should move to Backlog to revisit in 3.0.

**Comment by grahamehorner** [comment]
please try to be consistent with async naming conventions StartAsync() StopAsync() on Hosting IWebHost would suggest that interface naming should be IStartUpAsync and IStartupFilterAsync ? thoughts

**Comment by davidfowl** [comment]
I don't think that makes it any more consistent than `IAsyncStartupFilter`. The 2 things are completely unrelated.

**Comment by Tratcher** [comment]
Related: https://github.com/aspnet/Hosting/issues/373

**Comment by effyteva** [comment]
Any updates on this one? It's stuck for more than a year now..

**Comment by stijnherreman** [comment]
@effyteva see https://github.com/aspnet/Hosting/issues/1088#issuecomment-311741002, this issue will be revisited for the 3.0 release.
